File: CordialAppExtensions_Objective-C/CordialAppExtensions_Objective-C/CordialNotificationServiceExtension.m
//
//  CordialNotificationServiceExtension.m
//  CordialAppExtensions_Objective-C
//
//  Created by Yan Malinovsky on 7/30/19.
//

#import "CordialNotificationServiceExtension.h"

@interface CordialNotificationServiceExtension()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation CordialNotificationServiceExtension

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
    
    // Modify the notification content here...
    
    // Start
    
    NSString *urlString = request.content.userInfo[@"imageURL"];
    
    if (urlString == nil) {
        os_log(OS_LOG_DEFAULT, "CordialSDK_AppExtensions: Payload does not contain imageURL");
        self.contentHandler(self.bestAttemptContent);
    } else {
        os_log(OS_LOG_DEFAULT, "CordialSDK_AppExtensions: Payload contains imageURL");
    }
    
    NSURL *fileUrl = [[NSURL alloc] initWithString:urlString];
    
    if (fileUrl == nil) {
        os_log(OS_LOG_DEFAULT, "CordialSDK_AppExtensions: Error [imageURL isn’t a valid URL]");
        self.contentHandler(self.bestAttemptContent);
    }
    
    NSData *imageData = [[NSData alloc] initWithContentsOfURL:fileUrl];
    
    if (imageData == nil) {
        os_log(OS_LOG_DEFAULT, "CordialSDK_AppExtensions: Error downloading an image");
        self.contentHandler(self.bestAttemptContent);
    }
    
    NSFileManager *fileManager = NSFileManager.defaultManager;
    NSString *filePath = [NSTemporaryDirectory() stringByAppendingPathComponent:NSProcessInfo.processInfo.globallyUniqueString];
    NSURL *folderURL = [[NSURL alloc] initFileURLWithPath:filePath];
    
    BOOL success = NO;
    
    @try {
        NSError *error = nil;
        NSString *fileIdentifier = @"image.gif";
        [fileManager createDirectoryAtURL:folderURL withIntermediateDirectories:YES attributes:nil error:&error];
        NSURL *fileURL = [folderURL URLByAppendingPathComponent:fileIdentifier];
        [imageData writeToURL:fileURL options:NSDataWritingAtomic error:&error];
        UNNotificationAttachment *attachment = [UNNotificationAttachment attachmentWithIdentifier:fileIdentifier URL:fileURL options:nil error:&error];
        self.bestAttemptContent.attachments = [NSArray arrayWithObject:attachment];
        
        success = YES;
    } @catch (NSException *exception) {
        os_log(OS_LOG_DEFAULT, "CordialSDK_AppExtensions: Error [%{public}@]", exception.reason);
    } @finally {
        if (success) {
            os_log(OS_LOG_DEFAULT, "CordialSDK_AppExtensions: Image has been added successfully");
        } else {
            os_log(OS_LOG_DEFAULT, "CordialSDK_AppExtensions: Error saving an image");
        }
    }
    
    // End
    
    self.contentHandler(self.bestAttemptContent);
}

- (void)serviceExtensionTimeWillExpire {
    // Called just before the extension will be terminated by the system.
    // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
    os_log(OS_LOG_DEFAULT, "CordialSDK_AppExtensions: Image attaching failed by timeout");
    self.contentHandler(self.bestAttemptContent);
}

@end

File: Sources/CordialAppExtensions/NotificationServiceExtension/CordialNotificationServiceExtension.swift
//
//  CordialNotificationServiceExtension.swift
//  CordialAppExtensions
//
//  Created by Yan Malinovsky on 5/12/19.
//  Copyright © 2019 Cordial Experience, Inc. All rights reserved.
//

import Foundation
import UserNotifications
import os.log

open class CordialNotificationServiceExtension: UNNotificationServiceExtension {
    
    var contentHandler: ((UNNotificationContent) -> Void)?
    var bestAttemptContent: UNMutableNotificationContent?
    
    open override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
        self.contentHandler = contentHandler
        bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
        
        if let bestAttemptContent = bestAttemptContent {
            // Modify the notification content here...
            
            // Start
            
            var urlString:String? = nil
            if let imageURL = request.content.userInfo["imageURL"] as? String {
                os_log("CordialSDK_AppExtensions: Payload contains imageURL", log: .default, type: .info)
                urlString = imageURL
            } else {
                os_log("CordialSDK_AppExtensions: Payload does not contain imageURL", log: .default, type: .info)
            }
            
            if urlString != nil, let fileUrl = URL(string: urlString!) {
                guard let imageData = NSData(contentsOf: fileUrl) else {
                    os_log("CordialSDK_AppExtensions: Error downloading an image", log: .default, type: .error)
                    contentHandler(bestAttemptContent)
                    return
                }
                guard let attachment = UNNotificationAttachment.saveImageToDisk(fileIdentifier: "image.gif", data: imageData, options: nil) else {
                    os_log("CordialSDK_AppExtensions: Error saving an image", log: .default, type: .error)
                    contentHandler(bestAttemptContent)
                    return
                }
                
                bestAttemptContent.attachments = [ attachment ]
                
                os_log("CordialSDK_AppExtensions: Image has been added successfully", log: .default, type: .info)
            } else {
                os_log("CordialSDK_AppExtensions: Error [imageURL isn’t a valid URL]", log: .default, type: .error)
            }
            
            // End
            
            contentHandler(bestAttemptContent)
        }
    }
    
    open override func serviceExtensionTimeWillExpire() {
        // Called just before the extension will be terminated by the system.
        // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
        if let contentHandler = contentHandler, let bestAttemptContent =  bestAttemptContent {
            os_log("CordialSDK_AppExtensions: Image attaching failed by timeout", log: .default, type: .info)
            contentHandler(bestAttemptContent)
        }
    }
    
}

extension UNNotificationAttachment {
    
    static func saveImageToDisk(fileIdentifier: String, data: NSData, options: [NSObject : AnyObject]?) -> UNNotificationAttachment? {
        let fileManager = FileManager.default
        let folderName = ProcessInfo.processInfo.globallyUniqueString
        let folderURL = NSURL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(folderName, isDirectory: true)
        
        do {
            try fileManager.createDirectory(at: folderURL!, withIntermediateDirectories: true, attributes: nil)
            let fileURL = folderURL?.appendingPathComponent(fileIdentifier)
            try data.write(to: fileURL!, options: [])
            let attachment = try UNNotificationAttachment(identifier: fileIdentifier, url: fileURL!, options: options)
            return attachment
        } catch let error {
            os_log("CordialSDK_AppExtensions: Error [%{public}@]", log: .default, type: .error, error.localizedDescription)
        }
        
        return nil
    }
}
